### PR TITLE
[SPARK-30960][SQL] add back the legacy date/timestamp format support in CSV/JSON parser

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Before Spark 3.0, the JSON/CSV parser has a special behavior that, when the parser fails to parse a timestamp/date, fallback to another way to parse it, to support some legacy format. The fallback was removed by https://issues.apache.org/jira/browse/SPARK-26178 and https://issues.apache.org/jira/browse/SPARK-26243.

This PR adds back this legacy fallback. Since we switch the API to do datetime operations, we can't be exactly the same as before. Here we add back the support of the legacy formats that are common (examples of Spark 2.4):
1. the fields can have one or two letters
```
scala> sql("""select from_json('{"time":"1123-2-22 2:22:22"}', 'time Timestamp')""").show(false)
+-------------------------------------------+
|jsontostructs({"time":"1123-2-22 2:22:22"})|
+-------------------------------------------+
|[1123-02-22 02:22:22]                      |
+-------------------------------------------+
```
2. the separator between data and time can be "T" as well
```
scala> sql("""select from_json('{"time":"2000-12-12T12:12:12"}', 'time Timestamp')""").show(false)
+---------------------------------------------+
|jsontostructs({"time":"2000-12-12T12:12:12"})|
+---------------------------------------------+
|[2000-12-12 12:12:12]                        |
+---------------------------------------------+
```
3. the second fraction can be arbitrary length
```
scala> sql("""select from_json('{"time":"1123-02-22T02:22:22.123456789123"}', 'time Timestamp')""").show(false)
+----------------------------------------------------------+
|jsontostructs({"time":"1123-02-22T02:22:22.123456789123"})|
+----------------------------------------------------------+
|[1123-02-15 02:22:22.123]                                 |
+----------------------------------------------------------+
```
4. date string can end up with any chars after "T" or space
```
scala> sql("""select from_json('{"time":"1123-02-22Tabc"}', 'time date')""").show(false)
+----------------------------------------+
|jsontostructs({"time":"1123-02-22Tabc"})|
+----------------------------------------+
|[1123-02-22]                            |
+----------------------------------------+
```
5. remove "GMT" from the string before parsing
```
scala> sql("""select from_json('{"time":"GMT1123-2-22 2:22:22.123"}', 'time Timestamp')""").show(false)
+--------------------------------------------------+
|jsontostructs({"time":"GMT1123-2-22 2:22:22.123"})|
+--------------------------------------------------+
|[1123-02-22 02:22:22.123]                         |
+--------------------------------------------------+
```
### Why are the changes needed?

It doesn't hurt to keep this legacy support. It just makes the parsing more relaxed.

### Does this PR introduce any user-facing change?

yes, to make 3.0 support parsing most of the csv/json values that were supported before.

### How was this patch tested?

new tests

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala]: Unfortunately there is no UT for the json parser. The test-ability of the json parser is not good...

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: this was added when we remove the support

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: This test case was removed in 3.0. Now add it back.

**Comment by cloud-fan** [comment]
cc @MaxGekk @HyukjinKwon @srowen @xuanyuanking

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: Looking at what was removed there https://github.com/apache/spark/pull/23150/files#diff-c82e4b74d2a51fed29069745ce4f9e96L164 , `DateTimeUtils.stringToTime` and `DateTimeUtils.stringToTimestamp` are 2 different functions, see https://github.com/MaxGekk/spark/blob/60c0974261c947c0838457c40f4fe0e64d17ca15/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L173-L191 

`stringToTime` has a few problems - it doesn't respect to Spark's session time zone, and parses in the combined calendar. If you use it as a fallback, you can get parsed values in different calendars.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: Yea I just want to support the legacy format, not to be exactly the same with 2.4. We switch the calendar so we can't be exactly the same anyway.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: If we can't be exactly the same, why do we add this back?
cc @marmbrus

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: We can't be exactly the same in 1% cases, and this change can make us exactly the same in 99% cases.

Comment (marmbrus) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: - What cases used to be supported that are no longer supported?
 - What remains in the 1%?
 - Why can we not get 100%?

To me, Spark suddenly failing to parse timestamps that it used to parse fall into the "extremely annoying" category of regressions. Working with timestamps is already hard and confusing. If my job just suddenly starts returning `null` when it used to work as expected its:
 - Probably going to take a while for people to figure it out (while the `null` partition of their table mysteriously fills up).
 - Probably non-trival for them to figure out how to fix it. (I don't think constructing / testing format strings is a trivial activity).

I really question why we changed these API's rather than creating new better ones. These changes broke my pipeline and it took me hours + access to a team of spark committers to figure out why!!!!!

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: We switch the calendar from "hybrid calendar (Julian + Gregorian)" to "ISO chronology (Proleptic Gregorian)", which affects date/timestamp values before October 15, 1582. For example, some dates are invalid according to the new calendar but the old calendar accepts it. When we add days to a date, the calendar skips invalid dates.

The new calendar is widely used by RDBMS and others like parquet, snowflake, but we should have a better rollout plan for the calendar switch, like making the calendar configurable for Spark. Unfortunately, we catch this too late.

However, the more serious problem is the switch of the pattern string(as we switch the JDK datetime API internally), and 3.0 can't parse some date/timestamp strings suddenly. This is the first step to fix this issue by adding back the legacy fallback, we are still working on fixing the 1% when users set a custom pattern string:
1. try to support the previous pattern string with the new calendar. If can't, go 2)
2. detect the conflicts between old and new pattern string and throw an exception to tell users what to do (turn on the legacy config or accept the new pattern string knowing what it mean)

Comment (marmbrus) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: I agree with your analysis of the relative costs here.

I think there are probably very few Spark programs in existence where the specifics of date handling for dates before October 15, 1582 are critical to correctness. This is especially true if you still get back some valid date (rather than `null`). If thats really the only difference with the calendar switch, then I think you could even make an argument that making the calendar configurable is too high a cost for the project.

Removing support for previously supported date format parsing strings in contrast seems pretty disastrous to me. Having to remove test cases really should have been a red flag to reviewers of those changes that we were going to break users' programs.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: My five cents, besides of the calendar problem the fallback function `DateTimeUtils.stringToTime()` didn't take into account the session time zone (`spark.sql.session.timeZone`). This can lead to the situation when there are values in a column parsed in different time zones. I do think this is serious problem of the fallback mechanism. If we are going to restore it to only don't get nulls, this will make users life harder in troubleshooting wrong results.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: @MaxGekk now we call `DateTimeUtils.stringToTimestamp(str, options.zoneId)`, so timezone is not an issue?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: `stringToTimestamp` doesn't have the issue but you implemented different fallback mechanism. From my point of view, it makes sense to either:
- restore the old function `stringToTime`
- or keep current code as is.

Fallback to `stringToTimestamp` is new feature, I think.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: I know they are different implementations, but we care more about the behavior to end users.

Are there any timestamp strings that can be parsed by `stringToTime` but not `stringToTimestamp`?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: `stringToTimestamp` cannot parse time zone id with the `GMT` prefix, see https://github.com/MaxGekk/spark/blob/60c0974261c947c0838457c40f4fe0e64d17ca15/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L176

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: looks like a corner case, but let's add it back as well.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: @cloud-fan Here is a draft PR which supports any zone ids by `stringToTimestamp`: https://github.com/apache/spark/pull/27753

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: Thank you for keeping moving this situation!

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: @MaxGekk that's a good complement to support more legacy formats!

**Comment by gatorsmile** [comment]
cc @xuanyuanking

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala
Comment (xuanyuanking) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: Should we also shadow this fallback behavior under legacy config? Or JSON/CSV will not keep the same behavior with the SQL side?
As the current approach, it seems to break the rule we want to achieve in #27537: throw an exception when the result changing between old and new Spark versions.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: I don't think it should be protected by a config.

The fallback was there at the very beginning without any config, and I think it's reasonable to support the legacy format always, to make the parser more relaxed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: This will be fixed by #27753

Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Oh yeah, got it thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala]: Copy, that makes sense.
So this fallback logic is kind of guard logic for the parser, no matter the parser is new or legacy one. 
After this merged, #27537 need to address the logic conflict.

**Review by xuanyuanking — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.0.

**Comment by dongjoon-hyun** [comment]
Thank you all!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:260]: Should we rebase this?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:260]: good catch! I think we should.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:260]: Here is the PR https://github.com/apache/spark/pull/28453

**Pull request closed without merging**
